tests/cli/test_console.py

```diff
@@ -127,55 +235,93 @@ def test_msg_json_merge_list(self, output: TextIOWrapper):
+            pytest.param({}, {"isatty": False}, "^No output TTY available$", id="output-no-tty"),
+        ],
+        indirect=["stdin", "console_output"],
+    )
+    def test_ask_failure(
+        self,
+        monkeypatch: pytest.MonkeyPatch,
+        stdin: TextIOWrapper,
+        console_output: TextIOWrapper,
+        expected: str,
+    ):
         mock_input = Mock()
         monkeypatch.setattr("builtins.input", mock_input)
 
-        console = ConsoleOutput(output)
-        assert console.ask("test: ") is None
-        assert getvalue(output) == ""
+        console = ConsoleOutput(console_output=console_output)
+        user_input = ConsoleUserInputRequester(console)
+        with pytest.raises(OSError, match=expected):
+            user_input.ask("test")
         assert mock_input.call_args_list == []
 
-    def test_ask_no_stdin(self, monkeypatch: pytest.MonkeyPatch, output: TextIOWrapper):
-        monkeypatch.setattr("sys.stdin", None)
-
-        console = ConsoleOutput(output)
-        assert console.ask("test: ") is None
-        assert getvalue(output) == ""
-
-    def test_ask_input_exception(self, monkeypatch: pytest.MonkeyPatch, output: TextIOWrapper):
-        monkeypatch.setattr("builtins.input", Mock(side_effect=ValueError))
-
-        console = ConsoleOutput(output)
-        assert console.ask("test: ") is None
-        assert getvalue(output) == "test: "
-
-    def test_askpass(self, monkeypatch: pytest.MonkeyPatch, output: TextIOWrapper):
+    def test_ask_password(self, monkeypatch: pytest.MonkeyPatch, console_output: TextIOWrapper):
         def getpass(prompt, stream):
             stream.write(prompt)
             return "hello"
 
         monkeypatch.setattr("streamlink_cli.console.getpass", getpass)
 
-        console = ConsoleOutput(output)
-        assert console.askpass("test: ") == "hello"
-        assert getvalue(output) == "test: "
+        console = ConsoleOutput(console_output=console_output)
+        user_input = ConsoleUserInputRequester(console)
+        assert user_input.ask_password("test") == "hello"
+        assert getvalue(console_output) == "test: "
 
-    def test_askpass_no_tty(self, output: TextIOWrapper):
-        console = ConsoleOutput(output)
-        assert console.askpass("test: ") is None
-        assert getvalue(output) == ""
-
-    def test_askpass_no_stdin(self, monkeypatch: pytest.MonkeyPatch, output: TextIOWrapper):
-        monkeypatch.setattr("sys.stdin", None)
+    @pytest.mark.parametrize(
+        ("stdin", "console_output", "expected"),
+        [
+            pytest.param({"exists": False}, {}, "^No input TTY available$", id="no-stdin"),
+            pytest.param({"isatty": False}, {}, "^No input TTY available$", id="stdin-no-tty"),
+            pytest.param({}, {"exists": False}, "^No output TTY available$", id="no-output"),
+            pytest.param({}, {"isatty": False}, "^No output TTY available$", id="output-no-tty"),
```

Comment:
For consistency with the two `stdin` ids should the id values here be `no-stdout` and `stdout-no-tty`?

Reply:
`ConsoleOutput` can receive both `stdout` or `stderr`, depending on the used parameters and availability of `stdout`. Hence why it's internally also using the `console_output` name and why it's saying "output" on error.